src/proof_server.py

```diff
@@ -100,12 +110,92 @@ def handle_vote(self, sv_vote: sv_vote.SVVote):
         if sv_vote.proof_server_row is None:
             raise Exception('Cannot handle vote without valid proof_server_row')
         self._incoming_vote_rows[sv_vote.proof_server_row].append(sv_vote)
-
-    def publish_vote_consistency_proof(self, proof_lists: Set[int]) -> None:
+        
+    def _reverse_permutation(self, permutation: List[int], values: List[Any]):
+        back_tracked: List[Any] = [None for _ in values]
+        for i, val in enumerate(values):
+            back_tracked[permutation[i]] = val
+        return back_tracked
+    
+    def _unmix_commitments(self):
+        """
+        This function unmixes all final outputs, but does NOT de-obfuscate any of them.
+        """
+        commitment_arrays_unmixed = copy.deepcopy(self._commitment_arrays)
+        for list_idx in range(self._twoM):
+            for perm_array in reversed(self._permutation_arrays[list_idx]):
+                for row, row_svrs in enumerate(commitment_arrays_unmixed[list_idx]):
+                    un_mixed = self._reverse_permutation(perm_array, row_svrs)
+                    commitment_arrays_unmixed[list_idx][row] = un_mixed
+        self._unmixed_commitment_arrays = commitment_arrays_unmixed
+        
+    def _publish_t_values(self):
+        """
+        This also publishes the t-values per section II E, where u2 and v2 are the SV components
+        after obfuscation (in the mixing step) and u1 and v1 are the original components.
+        
+        This is part of step #7 from Section I and is briefly mentioned in Section IX
+        under "Proving consistency with cast votes"
         """
+        self._sbb.post_start_tvalue_commitments()
+        
+        t_values: List[List[List[Dict[str, int]]]]= []
+        for list_idx in range(self._twoM):
+            row_list = []
+            for row, final_svrs in enumerate(self._unmixed_commitment_arrays[list_idx]):
+                vote_list = []
+                initial_svrs = self._initial_sv[row]
+                assert len(initial_svrs) == len(final_svrs)
+                for vote_idx in range(len(initial_svrs)):
+                    initial_sv = initial_svrs[vote_idx]
+                    final_sv = final_svrs[vote_idx]
+                    tu = util.t_val(initial_sv.u, final_sv.u, self._M)
+                    tv = util.t_val(initial_sv.v, final_sv.v, self._M)
```

Comment:
In Section II-E, the t value for the v component is: `t = v_1 - v_2 (mod M)`, whereas here you are computing both t values the same (the way `u` is computed in the paper). Why is this? Does it not make a difference?

Reply:
I thought it was a typo in the paper, actually, or it might not matter (based on the rivest implementation). But holy crap, great eye from a code review standpoint